src/Pact/Analyze/Eval/Term.hs

```diff
@@ -369,9 +376,17 @@ evalTerm = \case
     obj@(Object fields) <- evalTermO objT
     validateWrite writeType schema obj
     sRk <- symRowKey <$> evalTerm rowKey
-    tableWritten tn .= true
+
+    thisRowExists <- use $ rowExists tn sRk
+    let writeSucceeds = case writeType of
+          Pact.Insert -> bnot thisRowExists
+          Pact.Write  -> true
+          Pact.Update -> thisRowExists
+    succeeds %= (&&& writeSucceeds)
+
+    tableWritten tn .= writeSucceeds
```

Comment:
It seems like we should probably keep this as `tableWritten tn .= true`:

So that we can talk about how the (1) code doing a write interacts with (2) tx success in a property. In this light, it seems like we should think about `tableWritten` as "code tries to do a write" rather than "write occurs only when the tx succeeds", keeping the two concepts separate.

Right now our property language assumes tx success so this is current moot to end-users, but we should probably assume that we will be exposing this extra power. And additionally our tests currently are capable of using this extra power.

Reply:
`table-written` implies that the write actually succeeded. If we changed to your behavior we'd at least want to change the name to `attempted-table-write` (which is unwieldy).

But logically they're the same (modulo successful transactions), so this change would have no user-visible effect (currently).

In a future world, where we have the more powerful prop language and users can speak of success and failure, we'd like to be able to distinguish between (a) written and (b) attempted-to-write-but-failed (attempted-to-write-and-succeeded is the same as (a)). In this case we'd want to have both `table-written` and `failed-table-write`, I think. So this implementation would stay and we would want to add additional logic for failed-table-write.

Reply:
I guess I don't understand the lack of symmetry between reads and writes (see `tableRead tn .= true`). If we have properties like "malicious user can read sensitive data" and "malicious user can write db", then the two should be treated the same with respect to tx-success semantics?

It seems that the two (read and write) should have similar behavior, regardless of whether we choose to couple or decouple them from tx success.

Reply:
Ah, I see that `writeSucceeds` doesn't look at whether the tx is succeeding, only whether this particular row exists or not.

Reply:
In that case, I think this logic makes sense then. Thanks for going through this with me